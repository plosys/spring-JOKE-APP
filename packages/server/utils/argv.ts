export interface Argv {
  // //////////////////////////////////////////////////////////////////////////////////////////// //
  // Logging                                                                                      //
  // //////////////////////////////////////////////////////////////////////////////////////////// //
  /**
   * Amount by which the logging verbosity is decreased from its default.
   */
  quiet: number;

  /**
   * Amount by which the logging verbosity is increased from its default.
   */
  verbose: number;

  // //////////////////////////////////////////////////////////////////////////////////////////// //
  // Hosting                                                                                      //
  // //////////////////////////////////////////////////////////////////////////////////////////// //
  /**
   * The hostname on which Appsemble Studio is served.
   */
  host: string;

  /**
   * The port on which Appsemble should be started.
   *
   * @default 9999
   */
  port: number;

  /**
   * This should be true of the HTTP proxy protocol is used.
   *
   * @default false
   */
  proxy: boolean;

  /**
   * The app secret. This is used for various things.
   */
  secret: string;

  /**
   * The key used for all the data that is encrypted using AES.
   */
  aesSecret: string;

  // //////////////////////////////////////////////////////////////////////////////////////////// //
  // DNS                                                                                          //
  // //////////////////////////////////////////////////////////////////////////////////////////// //
  /**
   * The strategy used to configure DNS.
   */
  appDomainStrategy: string;

  /**
   * The name of the Kubernetes server that’s serving Appsemble.
   */
  serviceName: string;

  /**
   * The exposed port of the Kubernetes server that’s serving Appsemble.
   */
  servicePort: string;

  /**
   * The hostname on which the Kubernetes API is available.
   *
   * @default 'kubernetes.default.svc'
   */
  kubernetesServiceHost: string;

  /**
   * The port on which the Kubernetes API is available.
   *
   * @default 443
   */
  kubernetesServicePort: number | string;

  /**
   * Kubernetes annotations to apply to the ingress as a JSON string.
   */
  ingressAnnotations: string;

  /**
   * The class name of the ingress.
   *
   * @default 'nginx'
   */
  ingressClassName: string;

  /**
   * The name of the cert-manager issuer to use for apps.
   */
  issuer?: string;

  /**
   * The name of the cert-manager cluster issuer to use for apps.
   */
  clusterIssuer?: string;

  // //////////////////////////////////////////////////////////////////////////////////////////// //
  // Database                                                                                     //
  // //////////////////////////////////////////////////////////////////////////////////////////// //
  /**
   * The hostname of the database.
   */
  databaseHost: string;

  /**
   * The port on which the database is accessed.
   */
  databasePort: number;

  /**
   * The username used to access the database.
   */
  databaseUser: string;

  /**
   * The password used to access the database.
   */
  databasePassword: string;

  /**
   * The name of the database to use.
   */
  databaseName: string;

  /**
   * Whether or not to use SSL for the database connection.
   *
   * @default false
   */
  databaseSsl: boolean;

  /**
   * The URL for the database. This replaces all other database options.
   */
  databaseUrl: string;

  /**
   * The version to migrate to.
   */
  migrateTo: string;

  // //////////////////////////////////////////////////////////////////////////////////////////// //
  // SSL                                                                                          //
  // //////////////////////////////////////////////////////////////////////////////////////////// //
  /**
   * Whether or not to use SSL
   *
   * @default false
   */
  ssl: boolea