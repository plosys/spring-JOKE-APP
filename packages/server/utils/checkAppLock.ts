
import { forbidden } from '@hapi/boom';
import { Context } from 'koa';

import { App } from '../models/index.js';

/**
 * Check if the app is currently locked.
 *
 * Will throw a 403 error if the app is locked.
 *
 * @param ctx The Koa context that can contain a force flag in its body
 * @param app The app to check against
 */
export function checkAppLock(ctx: Context, app: App): void {
  if (app.locked && !ctx.request.body?.force) {
    throw forbidden('App is currently locked.');
  }
}