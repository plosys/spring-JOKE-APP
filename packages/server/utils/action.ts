import { logger } from '@appsemble/node-utils';
import { defaultLocale, remap, RemapperContext } from '@appsemble/utils';

import { actions, ServerActionParameters } from './actions/index.js';
import { argv } from './argv.js';

export async function handleAction(
  action: (params: ServerActionParameters) => Promise<unknown>,
  params: ServerActionParameters,
): Promise<void> {
  logger.info(`Running action: ${params.action.type}`);
  const url = new URL(argv.host);
  url.hostname =
    params.app.domain || `${params.app.path}.${params.app.OrganizationId}.${url.hostname}`;
  const appUrl = String(url);
  const context: RemapperContext = params.internalContext ?? {
    appId: params.app.id,
    appUrl,
    url: String(url),
    context: {},
    history: [],
    // XXX: Implement getMessage and default language selections
    getMessage() {
      return null;
    },
    userInfo: undefined,
    locale: params.app.definition.defaultLanguage ?? defaultLocale,
  };
  let data =
    'remapBefore' in params.action
      ? remap(params.action.remapBefore, params.data, context)
      : params.data;

  const updatedContext = { ...context, history: [...context.history, data] };

  try {
    data = await action({ ...params, data, internalContext: updatedContext });
    if ('remapAfter' in params.action) {
      data = remap(params.action.remapAfter, data, updatedContext);
    }
    if (params.action.onSuccess) {
      await handleAction(actions[params.action.onSuccess.type], {
        ...params,
        action: params.action.onSuccess,
        data,
        internalContext: updatedContext,
      });
    }
  } catch (error) {
    logger.error(`Error running action: ${params.action.type}`);
    if (params.action.onError) {
      return handleAction(actions[params.action.onError.type], {
        ...params,
        action: params.action.onError,
        data,
        internalContext: updatedContext,
      });
    }
    throw error;
  }
  logger.info(`Successfully ran action: ${params.action.type}`);
}
