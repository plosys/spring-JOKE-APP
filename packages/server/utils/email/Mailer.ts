
import { logger } from '@appsemble/node-utils';
import { defaultLocale, has, IntlMessageFormat } from '@appsemble/utils';
import addrs, { ParsedMailbox } from 'email-addresses';
import { FormatXMLElementFn, PrimitiveType } from 'intl-messageformat';
import tags from 'language-tags';
import { createTransport, SendMailOptions as MailerSendMailOptions, Transporter } from 'nodemailer';
import { Options } from 'nodemailer/lib/smtp-transport';
import { Op } from 'sequelize';

import { App, AppMessages } from '../../models/index.js';
import { argv } from '../argv.js';
import { decrypt } from '../crypto.js';
import { getAppsembleMessages, getSupportedLanguages } from '../getAppsembleMessages.js';
import { readAsset } from '../readAsset.js';
import { renderEmail } from './renderEmail.js';

const supportedLanguages = getSupportedLanguages();
export interface Recipient {
  /**
   * The email address of the recipient.
   */
  email: string;

  /**
   * The name of the recipient.
   */
  name?: string;
}

export interface SendMailOptions {
  /**
   * The email address of the recipient
   */
  to?: string;

  /**
   * The name of the email sender.
   *
   * @default 'Appsemble'
   */
  from?: string;

  /**
   * The email address(es) to BCC the mail to.
   */
  cc?: string[] | string;

  /**
   * The email address(es) to BCC the mail to.
   */
  bcc?: string[] | string;

  /**
   * The subject of the email.
   */
  subject: string;

  /**
   * The HTML content of the email.
   */
  html: string;

  /**
   * The plain-text content of the email.
   */
  text: string;

  /**
   * The attachments to include in the email.
   */
  attachments?: MailerSendMailOptions['attachments'];

  /**
   * An app containing custom SMTP settings.
   */
  app?: Pick<
    App,
    'emailHost' | 'emailName' | 'emailPassword' | 'emailPort' | 'emailSecure' | 'emailUser'
  >;
}

/**
 * A class to simplify sending emails.
 */
export class Mailer {
  transport: Transporter;

  constructor({
    smtpFrom,
    smtpHost,
    smtpPass,
    smtpPort,
    smtpSecure,
    smtpUser,
  }: Pick<
    typeof argv,
    'smtpFrom' | 'smtpHost' | 'smtpPass' | 'smtpPort' | 'smtpSecure' | 'smtpUser'
  >) {
    if (smtpHost) {
      const auth = (smtpUser && smtpPass && { user: smtpUser, pass: smtpPass }) || null;
      this.transport = createTransport(
        {
          port: smtpPort || smtpSecure ? 465 : 587,
          pool: true,
          host: smtpHost,
          secure: smtpSecure,
          auth,
        } as Options,
        { from: smtpFrom },
      );
    }
  }

  /**
   * Check if the SMTP connection still works
   *
   * @throws If the SMTP connection no longer works.
   */
  async verify(): Promise<void> {
    if (!this.transport) {
      logger.warn('SMTP hasn’t been configured.');
      return;
    }
    await this.transport.verify();
  }

  async sendTranslatedEmail({
    app,
    appId,
    emailName,
    from = 'Appsemble',
    locale = defaultLocale,
    to,
    values,
  }: {
    to: Recipient;
    appId: number;
    from?: string;
    emailName: string;
    values: Record<string, FormatXMLElementFn<string, string[] | string> | PrimitiveType>;
    locale: string;
    app?: App;
  }): Promise<void> {
    const emailLocale = locale || defaultLocale;
    const lang = emailLocale.toLowerCase();
    const baseLanguage = tags(lang)
      .subtags()
      .find((sub) => sub.type() === 'language');
    const baseLang = baseLanguage && String(baseLanguage).toLowerCase();
    const appMessages = await AppMessages.findAll({
      where: {
        AppId: appId,
        language: { [Op.or]: baseLang ? [baseLang, lang, defaultLocale] : [lang, defaultLocale] },
      },
    });

    const subjectKey = `server.emails.${emailName}.subject`;
    const bodyKey = `server.emails.${emailName}.body`;

    let templateSubject: string;
    let templateBody: string;

    const langMessages = appMessages.find((a) => a.language === lang);
    const baseLangMessages = appMessages.find((a) => a.language === baseLang);
    const defaultLocaleMessages = appMessages.find((a) => a.language === defaultLocale);

    if (
      langMessages &&
      has(langMessages.messages?.core, subjectKey) &&
      has(langMessages.messages?.core, bodyKey)
    ) {
      templateSubject = langMessages.messages.core[subjectKey];
      templateBody = langMessages.messages.core[bodyKey];
    } else if (
      baseLangMessages &&
      has(baseLangMessages.messages?.core, subjectKey) &&
      has(baseLangMessages.messages?.core, bodyKey)
    ) {
      templateSubject = baseLangMessages.messages.core[subjectKey];
      templateBody = baseLangMessages.messages.core[bodyKey];
    } else if (
      defaultLocaleMessages &&
      has(defaultLocaleMessages.messages?.core, subjectKey) &&
      has(defaultLocaleMessages.messages?.core, bodyKey)
    ) {
      templateSubject = defaultLocaleMessages.messages.core[subjectKey];
      templateBody = defaultLocaleMessages.messages.core[bodyKey];
    } else if ((await supportedLanguages).has(baseLang) || (await supportedLanguages).has(lang)) {
      const coreMessages = await getAppsembleMessages(lang, baseLang);
      if (has(coreMessages, bodyKey) && has(coreMessages, subjectKey)) {
        templateSubject = coreMessages[subjectKey];
        templateBody = coreMessages[bodyKey];
      }
    }

    if (!templateSubject || !templateBody) {
      const messages = await getAppsembleMessages(defaultLocale);
      templateSubject = messages[subjectKey];
      templateBody = messages[bodyKey];
    }

    const sub = new IntlMessageFormat(templateSubject, emailLocale).format(values);
    const body = new IntlMessageFormat(templateBody, emailLocale).format(values);

    const { html, subject, text } = await renderEmail(body as string, {}, sub as string);

    const email = {
      to: to.name ? `${to.name} <${to.email}>` : to.email,
      from: from || 'Appsemble',
      subject,
      html,
      text,
    };

    if (app?.emailHost && app?.emailUser && app?.emailPassword) {
      const smtpPass = decrypt(app.emailPassword, argv.aesSecret);
      const mailer = new Mailer({
        smtpFrom: from,
        smtpHost: app.emailHost,
        smtpPass,
        smtpPort: app.emailPort,
        smtpSecure: app.emailSecure,
        smtpUser: app.emailUser,
      });

      await mailer.sendEmail(email);
      return;
    }

    await this.sendEmail(email);
  }

  /**
   * Send an email using the configured SMTP transport.
   *
   * @param to The email
   * @param templateName The name of the Markdown email template to send
   * @param values A key/value pair of values to use for rendering the email.
   */
  async sendTemplateEmail(
    to: Recipient,
    templateName: string,
    values: Record<string, string>,
  ): Promise<void> {
    const template = await readAsset(`email/${templateName}.md`, 'utf8');
    const { html, subject, text } = await renderEmail(template, {
      ...values,
      greeting: to.name ? `Hello ${to.name}` : 'Hello',
    });

    await this.sendEmail({
      to: to.name ? `${to.name} <${to.email}>` : to.email,
      subject,
      html,
      text,
    });
  }

  /**
   * Send an email using the configured SMTP transport.
   *
   * @param options The options specifying the contents and metadata of the email
   */
  async sendEmail({
    app,
    attachments = [],
    bcc,
    cc,
    from,
    html,
    subject,
    text,
    to,
  }: SendMailOptions): Promise<void> {
    let { transport } = this;
    if (app?.emailHost && app?.emailUser && app?.emailPassword) {
      const smtpPass = decrypt(app.emailPassword, argv.aesSecret);
      const mailer = new Mailer({
        smtpFrom: from,
        smtpHost: app.emailHost,
        smtpPass,
        smtpPort: app.emailPort,
        smtpSecure: app.emailSecure,
        smtpUser: app.emailUser,
      });

      ({ transport } = mailer);
    }

    if (!transport) {
      logger.warn('SMTP hasn’t been configured. Not sending real email.');
    }
    const parsed = addrs.parseOneAddress(argv.smtpFrom) as ParsedMailbox;
    const fromHeader = from ? `${from} <${parsed?.address}>` : argv.smtpFrom;

    const loggingMessage = ['Sending email:', `To: ${to}`];
    if (cc) {
      loggingMessage.push(`CC: ${cc}`);
    }
    if (bcc) {
      loggingMessage.push(`BCC: ${bcc}`);
    }
    if (fromHeader) {
      loggingMessage.push(`From: ${fromHeader}`);
    }
    loggingMessage.push(`Subject: ${subject}`, '', text);
    logger.info(loggingMessage.join('\n'));

    if (attachments.length) {
      logger.info(
        `Including ${attachments.length} attachments: ${JSON.stringify(
          attachments.map((a) => a.path || a.filename),
        )}`,
      );
    }
    if (transport) {
      await transport.sendMail({
        html,
        from: fromHeader,
        subject,
        text,
        to,
        attachments,
      });
    }
    logger.verbose('Email sent successfully.');
  }
}