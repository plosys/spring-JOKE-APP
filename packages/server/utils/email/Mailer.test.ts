
import { defaultLocale } from '@appsemble/utils';
import { setTestApp } from 'axios-test-instance';
import { Transporter } from 'nodemailer';

import { App, AppMessages, Organization } from '../../models/index.js';
import { argv, setArgv } from '../argv.js';
import { createServer } from '../createServer.js';
import { useTestDatabase } from '../test/testSchema.js';
import { Mailer } from './Mailer.js';

let mailer: Mailer;

useTestDatabase(import.meta);

beforeEach(() => {
  setArgv({ host: '', smtpFrom: 'test@example.com' });
  mailer = new Mailer(argv);
});

describe('verify', () => {
  it('should succeed if no transport exists', async () => {
    expect(await mailer.verify()).toBeUndefined();
  });

  it('should succeed if the transport verification succeeds', async () => {
    mailer.transport = {
      verify: () => Promise.resolve(true as const),
    } as Partial<Transporter> as Transporter;
    expect(await mailer.verify()).toBeUndefined();
  });

  it('should fail if the transport verification fails', async () => {
    mailer.transport = {
      verify: () => Promise.reject(new Error('fail')),
    } as Partial<Transporter> as Transporter;
    await expect(mailer.verify()).rejects.toThrow(new Error('fail'));
  });
});

describe('sendEmail', () => {
  it('should send emails with a name', async () => {
    mailer.transport = {
      sendMail: import.meta.jest.fn(() => null),
    } as Partial<Transporter> as Transporter;
    await mailer.sendTemplateEmail({ email: 'test@example.com', name: 'Me' }, 'resend', {
      url: 'https://example.appsemble.app/verify?code=test',
      name: 'Test App',
    });
    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      to: 'Me <test@example.com>',
      from: 'test@example.com',
      subject: 'Confirm account registration',
      text: expect.any(String),
      html: expect.any(String),
      attachments: [],
    });
  });

  it('should send emails without a name', async () => {
    mailer.transport = {
      sendMail: import.meta.jest.fn(() => null),
    } as Partial<Transporter> as Transporter;
    await mailer.sendTemplateEmail({ email: 'test@example.com' }, 'resend', {
      url: 'https://example.appsemble.app/verify?code=test',
      name: 'Test App',
    });
    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      to: 'test@example.com',
      from: 'test@example.com',
      subject: 'Confirm account registration',
      text: expect.any(String),
      html: expect.any(String),
      attachments: [],
    });
  });

  it('should not send emails when smtp is not configured', async () => {
    expect(
      await mailer.sendTemplateEmail({ email: 'test@example.com' }, 'resend', {
        url: 'https://example.appsemble.app/verify?code=test',
        name: 'The Appsemble Team',
      }),
    ).toBeUndefined();
  });
});

describe('sendTranslatedEmail', () => {
  let app: App;
  const supportedLocales = [defaultLocale, 'nl'];

  const tests = {
    resend: [
      {
        name: 'null',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
      {
        name: 'John Doe',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
    ],
    reset: [
      {
        name: 'null',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
      {
        name: 'John Doe',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
    ],
    welcome: [
      {
        name: 'null',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
      {
        name: 'John Doe',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
    ],
    appMemberEmailChange: [
      {
        name: 'null',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
      {
        name: 'John Doe',
        link: (text: string) => `[${text}](https://example.com)`,
        appName: 'Test App',
      },
    ],
  };

  beforeEach(() => {
    mailer.transport = {
      sendMail: import.meta.jest.fn(),
    } as Partial<Transporter> as Transporter;
  });

  beforeEach(async () => {
    const server = await createServer();
    await setTestApp(server);
    const organization = await Organization.create({
      id: 'testorganization',
      name: 'Test Organization',
    });
    app = await App.create({
      definition: {
        name: 'Test App',
        defaultPage: 'Test Page',
        security: {
          default: {
            role: 'Reader',
            policy: 'everyone',
          },
          roles: {
            Reader: {},
            Admin: {},
          },
        },
      },
      path: 'test-app',
      vapidPublicKey: 'a',
      vapidPrivateKey: 'b',
      OrganizationId: organization.id,
    });
  });

  it('should send emails in the default language', async () => {
    await mailer.sendTranslatedEmail({
      appId: app.id,
      emailName: 'welcome',
      to: { email: 'test@example.com', name: 'John Doe' },
      locale: 'en',
      values: {
        name: 'John Doe',
        appName: 'Test App',
        link: (text) => `[${text}](http://example.com/token=abcdefg)`,
      },
    });

    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      attachments: [],
      from: 'Appsemble <test@example.com>',
      html: `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello John Doe,</p>
<p>Thank you for registering your account. Before you can use your account, we need to verify your email address.</p>
<p>Please click <a href="http://example.com/token=abcdefg">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>Test App</em></p>
</body>
</html>
`,
      subject: 'Welcome to Test App',
      text: `Hello John Doe,

Thank you for registering your account. Before you can use your account, we need to verify your email address.

Please click [here](http://example.com/token=abcdefg) to verify your email address.

Kind regards,

_Test App_
`,
      to: 'John Doe <test@example.com>',
    });
  });

  it('should send emails in another default supported language', async () => {
    await mailer.sendTranslatedEmail({
      appId: app.id,
      emailName: 'welcome',
      to: { email: 'test@example.com', name: 'John Doe' },
      locale: 'nl',
      values: {
        name: 'John Doe',
        appName: 'Test App',
        link: (text) => `[${text}](http://example.com/token=abcdefg)`,
      },
    });

    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      attachments: [],
      from: 'Appsemble <test@example.com>',
      html: `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Beste John Doe,</p>
<p>Bedankt voor het registeren van jouw account. Voordat je jouw account kan gebruiken, moeten we jouw e-mailadres verifiëren.</p>
<p>Klik <a href="http://example.com/token=abcdefg">hier</a> om jouw e-mailadres te verifiëren.</p>
<p>Met vriendelijke groet,</p>
<p><em>Test App</em></p>
</body>
</html>
`,
      subject: 'Welkom bij Test App',
      text: `Beste John Doe,

Bedankt voor het registeren van jouw account. Voordat je jouw account kan gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](http://example.com/token=abcdefg) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_Test App_
`,
      to: 'John Doe <test@example.com>',
    });
  });

  it('should use fall back to the english translations if an app’s email translations don’t exist', async () => {
    await mailer.sendTranslatedEmail({
      appId: app.id,
      emailName: 'welcome',
      to: { email: 'test@example.com', name: 'John Doe' },
      locale: 'jp',
      values: {
        name: 'John Doe',
        appName: 'Test App',
        link: (text) => `[${text}](http://example.com/token=abcdefg)`,
      },
    });

    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      attachments: [],
      from: 'Appsemble <test@example.com>',
      html: `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello John Doe,</p>
<p>Thank you for registering your account. Before you can use your account, we need to verify your email address.</p>
<p>Please click <a href="http://example.com/token=abcdefg">here</a> to verify your email address.</p>
<p>Kind regards,</p>
<p><em>Test App</em></p>
</body>
</html>
`,
      subject: 'Welcome to Test App',
      text: `Hello John Doe,

Thank you for registering your account. Before you can use your account, we need to verify your email address.

Please click [here](http://example.com/token=abcdefg) to verify your email address.

Kind regards,

_Test App_
`,
      to: 'John Doe <test@example.com>',
    });
  });

  it('should use an app’s email translations', async () => {
    await AppMessages.create({
      AppId: app.id,
      language: 'nl-nl',
      messages: {
        core: {
          'server.emails.welcome.subject': 'Aangenaam!',
          'server.emails.welcome.body': 'Hoi {name}!',
        },
      },
    });
    await mailer.sendTranslatedEmail({
      appId: app.id,
      emailName: 'welcome',
      to: { email: 'test@example.com', name: 'John Doe' },
      locale: 'nl-nl',
      values: {
        name: 'John Doe',
        appName: 'Test App',
        link: (text) => `[${text}](http://example.com/token=abcdefg)`,
      },
    });

    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      attachments: [],
      from: 'Appsemble <test@example.com>',
      html: `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hoi John Doe!</p>
</body>
</html>
`,
      subject: 'Aangenaam!',
      text: 'Hoi John Doe!\n',
      to: 'John Doe <test@example.com>',
    });
  });

  it('should use an app’s email translations for the base language if the selected locale isn’t directly translated', async () => {
    await AppMessages.create({
      AppId: app.id,
      language: 'nl',
      messages: {
        core: {
          'server.emails.welcome.subject': 'Aangenaam!',
          'server.emails.welcome.body': 'Hoi {name}!',
        },
      },
    });
    await mailer.sendTranslatedEmail({
      appId: app.id,
      emailName: 'welcome',
      to: { email: 'test@example.com', name: 'John Doe' },
      locale: 'nl-nl',
      values: {
        name: 'John Doe',
        appName: 'Test App',
        link: (text) => `[${text}](http://example.com/token=abcdefg)`,
      },
    });

    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      attachments: [],
      from: 'Appsemble <test@example.com>',
      html: `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hoi John Doe!</p>
</body>
</html>
`,
      subject: 'Aangenaam!',
      text: 'Hoi John Doe!\n',
      to: 'John Doe <test@example.com>',
    });
  });

  it('should use an app’s english email translation overrides if the base language if the selected locale isn’t translated', async () => {
    await AppMessages.create({
      AppId: app.id,
      language: 'en',
      messages: {
        core: {
          'server.emails.welcome.subject': 'Hello!',
          'server.emails.welcome.body': 'How do you do, {name}?',
        },
      },
    });
    await mailer.sendTranslatedEmail({
      appId: app.id,
      emailName: 'welcome',
      to: { email: 'test@example.com', name: 'John Doe' },
      locale: 'jp',
      values: {
        name: 'John Doe',
        appName: 'Test App',
        link: (text) => `[${text}](http://example.com/token=abcdefg)`,
      },
    });

    expect(mailer.transport.sendMail).toHaveBeenCalledWith({
      attachments: [],
      from: 'Appsemble <test@example.com>',
      html: `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>How do you do, John Doe?</p>
</body>
</html>
`,
      subject: 'Hello!',
      text: 'How do you do, John Doe?\n',
      to: 'John Doe <test@example.com>',
    });
  });

  describe.each(supportedLocales)('%s', (locale) => {
    it('should support %s', () => {
      expect(supportedLocales.includes(locale)).toBeTruthy();
    });
  });

  describe.each(Object.entries(tests))('%s', (name, testValues) => {
    describe.each(supportedLocales)('%s', (locale) => {
      it.each(testValues)(`should render ${name} %# for locale ${locale}`, async (values) => {
        await mailer.sendTranslatedEmail({
          appId: app.id,
          emailName: name,
          to: { email: 'test@example.com', name: values.name },
          values: values as any,
          locale,
        });

        expect(mailer.transport.sendMail).toHaveBeenCalledTimes(1);
        // The subject
        expect((mailer.transport.sendMail as jest.Mock).mock.calls[0][2]).toMatchSnapshot();
        // The body
        expect((mailer.transport.sendMail as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
});