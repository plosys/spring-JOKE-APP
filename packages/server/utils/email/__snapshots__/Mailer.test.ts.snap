
// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange 0 for locale en 1`] = `undefined`;

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange 0 for locale en 2`] = `
{
  "attachments": [],
  "from": "Appsemble <test@example.com>",
  "html": "<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>You have changed your email address for Test App. Before you can use your account, we need to verify your email address.</p>
<p>Please click <a href="https://example.com">here</a> to verify your email address.</p>
<p>Kind regards</p>
<p>, <em>Test App</em></p>
</body>
</html>
",
  "subject": "Email changed",
  "text": "Hello,

You have changed your email address for Test App. Before you can use your account, we need to verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards

, _Test App_
",
  "to": "null <test@example.com>",
}
`;

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange 1 for locale en 1`] = `undefined`;

exports[`sendTranslatedEmail appMemberEmailChange en should render appMemberEmailChange 1 for locale en 2`] = `
{
  "attachments": [],
  "from": "Appsemble <test@example.com>",
  "html": "<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello John Doe,</p>
<p>You have changed your email address for Test App. Before you can use your account, we need to verify your email address.</p>
<p>Please click <a href="https://example.com">here</a> to verify your email address.</p>
<p>Kind regards</p>
<p>, <em>Test App</em></p>
</body>
</html>
",
  "subject": "Email changed",
  "text": "Hello John Doe,

You have changed your email address for Test App. Before you can use your account, we need to verify your email address.

Please click [here](https://example.com) to verify your email address.

Kind regards

, _Test App_
",
  "to": "John Doe <test@example.com>",
}
`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange 0 for locale nl 1`] = `undefined`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange 0 for locale nl 2`] = `
{
  "attachments": [],
  "from": "Appsemble <test@example.com>",
  "html": "<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hallo,</p>
<p>Je hebt je e-mailadres voor Test App gewijzigd. Voordat je jouw account kunt gebruiken, moeten we jouw e-mailadres verifiëren.</p>
<p>Klik <a href="https://example.com">hier</a> om jouw e-mailadres te verifiëren.</p>
<p>Met vriendelijke groet,</p>
<p><em>Test App</em></p>
</body>
</html>
",
  "subject": "E-mail gewijzigd",
  "text": "Hallo,

Je hebt je e-mailadres voor Test App gewijzigd. Voordat je jouw account kunt gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_Test App_
",
  "to": "null <test@example.com>",
}
`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange 1 for locale nl 1`] = `undefined`;

exports[`sendTranslatedEmail appMemberEmailChange nl should render appMemberEmailChange 1 for locale nl 2`] = `
{
  "attachments": [],
  "from": "Appsemble <test@example.com>",
  "html": "<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Beste John Doe,</p>
<p>Je hebt je e-mailadres voor Test App gewijzigd. Voordat je jouw account kunt gebruiken, moeten we jouw e-mailadres verifiëren.</p>
<p>Klik <a href="https://example.com">hier</a> om jouw e-mailadres te verifiëren.</p>
<p>Met vriendelijke groet,</p>
<p><em>Test App</em></p>
</body>
</html>
",
  "subject": "E-mail gewijzigd",
  "text": "Beste John Doe,

Je hebt je e-mailadres voor Test App gewijzigd. Voordat je jouw account kunt gebruiken, moeten we jouw e-mailadres verifiëren.

Klik [hier](https://example.com) om jouw e-mailadres te verifiëren.

Met vriendelijke groet,

_Test App_
",
  "to": "John Doe <test@example.com>",
}
`;

exports[`sendTranslatedEmail resend en should render resend 0 for locale en 1`] = `undefined`;

exports[`sendTranslatedEmail resend en should render resend 0 for locale en 2`] = `
{
  "attachments": [],
  "from": "Appsemble <test@example.com>",
  "html": "<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
<p>Hello,</p>
<p>Someone has requested to resend the verification key for your account. Before you can use your account, we need to verify your email address.</p>
<p>Please click <a href="https://example.com">this link</a> to vervify your email address.</p>
<p><strong>Didn’t request this email?</strong></p>
<p>No worries! Your address may have been entered by mistake. If you ignore or delete this email, nothing will happen to your account.</p>
<p>Kind regards,</p>
<p><em>Test App</em></p>
</body>
</html>
",
  "subject": "Confirm account registration",
  "text": "Hello,